### ISSUE TITLE
Flatpak package

### ISSUE THREAD

**Issue by ghost** [input-leap/input-leap#30]
> This issue has been migrated from old Barrier Github repository https://github.com/debauchee/barrier/issues/30
> 
> Issue created on: 2018-04-21 by @AdrianKoshka
> Issue last updated on: 2018-12-31

I was wondering if there'd be any interest in packaging barrier in flatpak format. Distros like [Fedora Atomic Workstation](https://www.projectatomic.io/blog/2018/02/fedora-atomic-workstation/) encourage the use of flatpaks for GUI applications over traditionally packaged applications. I help maintain the Thunderbird flatpak so I'd be willing to help package barrier. Getting it up on flathub would also help.

---

> Commented on: 2018-04-22 by @walker0643

hi @AdrianKoshka - of course we would have no objection to that. if you're offering to maintain the package then just let us know what you need from us to get started. cheers!

---

> Commented on: 2018-04-22 by @AdrianKoshka

A list of dependencies that barrier needs to compile against would be a good start, and then how to build in on linux. I see there's an issue filed already for build instructions for linux (#28).

---

> Commented on: 2018-04-22 by @AdrianKoshka

Seems avahi libraries aren't part of the KDE/FreeDesktop runtime, what version of avahi is barrier 2.0.0 built against?

---

> Commented on: 2018-04-24 by @AdrianKoshka

If you want to keep up with the repo: https://github.com/AdrianKoshka/flathub/tree/com.gituhb.debauchee.barrier

I don't think I'll make the PR to flathub until I can at least get this thing to at least build.

---

> Commented on: 2018-04-24 by @AdrianKoshka

> Seems avahi libraries aren't part of the KDE/FreeDesktop runtime, what version of avahi is barrier 2.0.0 built against?

In what capacity does barrier make use of avahi? I ask because I realize if I have to compile avahi daemon and all, it would just be more efficient to (in my opinion):

~- [x] Ask flatpak/xdg-portal to add Avahi protal~ (We just need to open the sandbox so it can be talked to over DBus)
- [ ] Only build dev libraries so barrier can at least build
- [ ] Go from there?

---

> Commented on: 2018-04-24 by @walker0643

avahi is used to do auto configuration (auto-connect) between server and client on the LAN. I've been linking it against 0.7 on linux platforms.

as a quick rundown, barrier also needs the following to build: CMake, Python, GNU Make, GCC v4.9+, X11 headers and libs, Qt 5 (core, ui, widgets, network, qmake), libcurl, and openssl headers and libraries.

there is a clean_build.sh script in the base directory to automate the process a little. anything you need to tweak in the environment should go into a new script beside it called build_env.sh (make it +x). git won't track it nor overwrite it when you pull (it's a site config).

---

> Commented on: 2018-04-24 by @AdrianKoshka

Thanks for the info.

---

> Commented on: 2018-04-24 by @AdrianKoshka

Do you know anything about building avahi (or it's libraries rather)?

---

> Commented on: 2018-04-25 by @walker0643

I've not built it yet; so far the target platforms have all had usable packages. Are you running into trouble?

---

> Commented on: 2018-04-25 by @AdrianKoshka

Yeah...the build options are just poorly documented or just not documented at all. I've disabled quite a bit but can't get everything I need disabled (like the daemon component).

---

> Commented on: 2018-04-26 by @AdrianKoshka

Thanks to @lathiat, I was able to get avahi to at least compile. Though still working out some stuff.

---

> Commented on: 2018-04-26 by @AdrianKoshka

So If I remember correctly the default config file is `~/.barrier.config`? right? I was wondering if instead you could use `XDG_CONFIG_HOME`. (so something like `~/.config/barrier/<files-within>`)

---

> Commented on: 2018-04-26 by @AdrianKoshka

Also use `XDG_DATA_HOME` to store the SSL certs?

---

> Commented on: 2018-04-26 by @AdrianKoshka

http://docs.flatpak.org/en/latest/conventions.html#xdg-base-directories

---

> Commented on: 2018-04-26 by @AdrianKoshka

Also, apparently the `.desktop` file's executable path is hardocded. An excerpt of a conversation I had with a flatpak dev on what's wrong with the `.desktop` file you supply with barrier:

```
TingPing: hardcoding /usr/bin is wrong any time you have a different prefix
TingPing: either A) Use just the binary name, no path. B) At configure time insert the correct prefix
```

---

> Commented on: 2018-04-26 by @AdrianKoshka

Also, there seems to be something wrong the `include` check you have CMake do. As of now, I have to manually tell it where to look for the `dns_sd.h` header for avahi in the flatpak manifest. An excerpt from the manifest:

```json
"build-options": {
    "cflags": "-O2 -g -I/app/include/avahi-compat-libdns_sd",
    "cxxflags": "-O2 -g -I/app/include/avahi-compat-libdns_sd",
    "env": {
      "V": "1"
    }
  },
```

[the full manifest](https://github.com/AdrianKoshka/flathub/blob/com.gituhb.debauchee.barrier/com.github.debauchee.barrier.json)

---

> Commented on: 2018-04-26 by @walker0643

setting BONJOUR_SDK_HOME in the environment should take care of the non-standard location of the avahi package. eg. if the header is /opt/bonjour/include/dns_sd.h and the lib is /opt/bonjour/lib/dnssd.lib then add "export BONJOUR_SDK_HOME=/opt/bonjour" to your build_env.sh file in the root of the source.

---

> Commented on: 2018-04-26 by @walker0643

as far as the config file locations, can you create a new issue for that so it can be addressed separately?

---

> Commented on: 2018-04-26 by @AdrianKoshka

Can do.

---

> Commented on: 2018-04-26 by @AdrianKoshka

Issue #31 created.

---

> Commented on: 2018-04-26 by @AdrianKoshka

> then add "export BONJOUR_SDK_HOME=/opt/bonjour" to your build_env.sh file in the root of the source.

The build scripts are actually bypassed entirely when I build with flatpak, I think flatpak-build just knows magically how to run CMake and friends. Though, I can probably add `BONJOUR_SDK_HOME` in the build env section of the manifest.

---

> Commented on: 2018-04-26 by @AdrianKoshka

```json
"build-options": {
        "env": {
          "BARRIER_REVISION": "2.0.0",
	      "BONJOUR_SDK_HOME": "/app/include"
        }
      },
```
Putting it in the barrier module section of the build manifest doesn't work.

---

> Commented on: 2018-04-26 by @AdrianKoshka

Setting it in the higher-level section of the build-manifest via:

```json
  "build-options": {
    "cflags": "-O2 -g",
    "cxxflags": "-O2 -g",
    "env": {
      "V": "1",
      "BONJOUR_SDK_HOME": "/app/include"
```
doesn't work either...

---

> Commented on: 2018-04-26 by @AdrianKoshka

The contents of `app/include` btw:

```
total 0
drwxrwxr-x. 2 alc alc  55 Apr 26 18:49 avahi-client
drwxrwxr-x. 2 alc alc 260 Apr 26 18:49 avahi-common
drwxrwxr-x. 2 alc alc  22 Apr 26 18:49 avahi-compat-libdns_sd
drwxrwxr-x. 2 alc alc  78 Apr 26 18:49 avahi-core
```

---

> Commented on: 2018-04-27 by @AdrianKoshka

This appears to be a CMake issue (or rather, an issue with your build files)?

because:

1. The env variable works just fine in a shell module
2. I was told so by somebody helping me in `#flatpak`

```
AdrianCeleste TingPing: I got /app/include from my shell module
TingPing then its just a cmake problem
AdrianCeleste So do I need to go back to using the -I option in the build-options section?
TingPing it should be fixed in the build files
AdrianCeleste Ah, so I need to tell upstream?
TingPing yes, they clearly knew it was a problem since they had freebsd workarounds, they expect everythign to be in /usr
AdrianCeleste :T Ah, right...
```

---

> Commented on: 2018-04-27 by @walker0643

oops that was my bad... BONJOUR_SDK_HOME only works on windows. sorry :stuck_out_tongue:

yes, if cmake isn't finding your bonjour libs then just add your include and link options as you were doing. it's certainly cleaner to find out why cmake isn't finding bonjour and fix that instead, but for now just getting through the build might be enough.

btw, might it have something to do with the fact that bonjour is under /app ?? i'm not familiar with flatpak but /app seems like a very non-standard location. is everything used by the build process under /app or just bonjour?

---

> Commented on: 2018-04-27 by @AdrianKoshka

Everything, kinda, I can get you a link of flatpak file-structure if you want. While I would rather not do that include thing, but if it makes it work for now and will actually get solved later, I'll do it. I haven't even gotten around to see if barrier launches/work yet, but will probably later today.

here's that [link](http://docs.flatpak.org/en/latest/conventions.html#filesystem-layout).

---

> Commented on: 2018-04-27 by @AdrianKoshka

I actually probably won't test that it works until #31 is solved, I don't feel like opening the sandbox.

---

> Commented on: 2018-04-28 by @AdrianKoshka

I lied, and I can confirm it at least launches. Oh, question, does Barrier need DRI access? I ask because I see this:

```
alc@am1m-s2h ~/git/github/flathub/flathub (git)-[com.gituhb.debauchee.barrier] % flatpak run com.github.debauchee.barrier
libGL error: MESA-LOADER: failed to retrieve device information
libGL error: unable to load driver: amdgpu_dri.so
libGL error: driver pointer missing
libGL error: failed to load driver: amdgpu
libGL error: failed to open drm device: No such file or directory
libGL error: failed to load driver: radeonsi
```

---

> Commented on: 2018-04-28 by @AdrianKoshka

![picture of barrier running via flatpak run](https://imgur.com/4yno90B.png)

---

> Commented on: 2018-04-28 by @AdrianKoshka

I added `--device=dri` (enables OpenGL rendering) to the `finish-args` section of the manifest, and it no-longer complains about the DRI stuff.

---

> Commented on: 2018-04-28 by @AdrianKoshka

Did a single test, and the keyboard/mouse sharing worked, as well as copy/paste:

- Host: Fedora 27, flatpak'd `barrier` (running without SSL [The screenshot above is from before I started testing]) 2.0.0
- Client: Debian 9 stretch, `~/.local/bin/barrierc --no-daemon <server-ip>` 2.0.0

I might do a test with SSL later.

---

> Commented on: 2018-04-28 by @walker0643

That's good news, Adrian. Cheers for sticking with it. Barrier itself doesn't explicitly access the dri device, though Qt5 might use it if it's available.

On a side note, all the cool kids use i3 :smile: 

---

> Commented on: 2018-04-28 by @AdrianKoshka

I guess what needs to be done now is:

- [ ] Write an appstream file
- [X] take care of #31 
- [ ] Some more testing
- [ ] PR to flathub?

---

> Commented on: 2018-04-29 by @AdrianKoshka

[Here's a rough-draft of the appstream metadta](https://github.com/AdrianKoshka/flathub/blob/com.github.debauchee.barrier/com.github.debauchee.barrier.appdata.xml)

---

> Commented on: 2018-04-29 by @AdrianKoshka

I built barrier as a 32-bit flatpak, and it runs! (flathub prefers if something can build for `i386`, `x86_64`, `arm`, and `aarch64`. I'd test `arm`/`aarch64` but I don't feel like spinning up VMs.) If you're wondering how I did it, I simply did:

`make BUILDER_OPTIONS="--arch=i386"` which passes that argument to `flatpak-builder` via a variable.

![32-bit barrier flatpak running](https://imgur.com/c0NWQVu.png)

---

> Commented on: 2018-04-29 by @AdrianKoshka

Here's the 32-bit flatpak'd barrier running on 32-bit `flatpak` on my netbook (64-bit kernel, mostly 32-bit userland, Debian 9 stretch [x86 raspbian spin]):

![32-bit flatpak netbook](https://imgur.com/wWajhsM.png)

---

> Commented on: 2018-05-22 by @AdrianKoshka

Apologies for the radio silence, I was on a vacation for the past two weeks, and went to VCF-E (Vintage Computer Festival East) this past weekend.

---

> Commented on: 2018-05-22 by @AdrianKoshka

Thanks for b43581c2f54ebb38eefe635352a8ae4ac980407e, I was able to [remove a section from my manifest](https://github.com/AdrianKoshka/flathub/commit/e74fd81b9cc27a31d3e067077e6766e8253316a7).

---

> Commented on: 2018-05-23 by @johnny-mac

So this is in active development on flatpak? that's awesome! But your link up there didn't work, page not found.. You gonna be adding it to flathub any time soon? 
I run Void Linux (GLIBC & MUSL) but can't get Barrier to build, would really love to be able to use a flatpak app for it.

---

> Commented on: 2018-05-23 by @AdrianKoshka

I'm going to PR to flathub when I feel it's ready. I already have one WIP PR to flathub (for virt-viewer/remote-viewer) so I don't have two WIP PR's.

If you want my repo so you can build from source: https://github.com/AdrianKoshka/flathub/tree/com.github.debauchee.barrier

---

> Commented on: 2018-05-23 by @AdrianKoshka

So far, the easiest way to build the flatpak is:

```shell
$ flatpak --user install flathub org.flatpak.Builder
$ git clone https://github.com/AdrianKoshka/flathub.git -b com.github.debauchee.barrier
$ cd flathub
$ make FLATPAK_BUILDER="flatpak run org.flatpak.Builder
```
Then to install locally:

```shell
$ flatpak --user remote-add --no-gpg-verify lbr /full/path/to/repo/
$ flatpak --user install lbr com.github.debauchee.barrier
```

---

> Commented on: 2018-06-30 by @AdrianKoshka

Why was this closed? Inactivity, I've been waiting on a PR you pulled into a different branch to be pulled into master. #48 specifically.

---

> Commented on: 2018-07-03 by @AdrianKoshka

Now that the `pkg-config` fixe was merged (#86), I can say that I no longer have to use the hack to point the compiler to the avahi libraries.

---

> Commented on: 2018-07-03 by @AdrianKoshka

This means I can finally start moving forward again, probably start working on the appstream data.

---

> Commented on: 2018-07-03 by @AdrianKoshka

How do you properly pass the version of barrier to the buildsystem? I have:

```json
{
            "name" : "barrier",
            "buildsystem" : "cmake-ninja",
            "build-options" : {
                "env" : {
                    "BARRIER_REVISION" : "2.1.2"
                }
            },
            "sources" : [
                {
                    "type" : "git",
                    "url" : "git://github.com/debauchee/barrier.git",
                    "tag" : "v2.1.2",
                    "commit" : "cc69299ea39bd39caad54ae349564e22593eb4df"
                }
```

Though when I run `--version`, I get:

```
alc@am1m-s2h ~ % flatpak run --command=barrierc com.github.debauchee.barrier --version
barrierc 2.1.0-snapshot
Protocol version 1.6
Copyright (C) 2018 Debauchee Open Source Group
Copyright (C) 2012-2016 Symless Ltd.
Copyright (C) 2008-2014 Nick Bolton
Copyright (C) 2002-2014 Chris Schoeneman
```

---

> Commented on: 2018-09-08 by @walker0643

The top of cmake/Version.cmake has the constants that cmake uses to fill in the placeholder variables like @BARRIER_VERSION@ in the .in files. Is that what you were asking??

---

> Commented on: 2018-12-31 by @ugtar

is avahi required? i personally don't have the avahi daemon running because i don't need to auto-discover anything.  With avahi not running, barrier can't auto-configure, but it works fine if I manually enter the barrier server IP.  Seems like avahi could be an optional dependency that just disables the auto-discovery option at build time.

---

> Commented on: 2018-12-31 by @AdrianKoshka

I'm not going to take avahi out just because one user doesn't use it. Also I should've closed this issue a while ago.

---

> Commented on: 2018-12-31 by @AdrianKoshka

@ugtar If your intention is to make avahi an optional build component (which it [seems to be, partially](https://github.com/debauchee/barrier/blob/master/CMakeLists.txt#L199-L201)), then open another issue instead of piggy-backing off of this one.

---

> Closed on: 2018-12-31 by @AdrianKoshka
